### PR TITLE
post data to paths/*/item/*

### PR DESCRIPTION
**Pull request by muriloogps**
I am using FormData, but since I am not very familiar with it, I would like you to take a look!

### CODE REVIEW

File: src/main/webapp/script.js [hunk 1]
```diff
@@ -17,12 +17,43 @@ $('#myModal').on('shown.bs.modal', function () {
...
 function submitFeedback() {
-  // TODO
-}
+    console.log("Sending the submission form info.");
+    const itemId = document.getElementById("feedback-id").innerText;
+    const pathId = document.getElementById("path-id").innerText;
+    const options = document.getElementById("ratingValues");
+    const rating = options.options[options.selectedIndex].value;
+
+    const XHR = new XMLHttpRequest();
```
Comment (Tochukwuibe) [inline src/main/webapp/script.js]: Is there a reason you chose to go with XHR over Fetch?

Comment (muriloogps) [reply src/main/webapp/script.js]: Not really. I can change it.

Comment (Tochukwuibe) [reply src/main/webapp/script.js]: If it's not too much to do yea that'd be cool,  fetch is the more recent solution.

Comment (muriloogps) [reply src/main/webapp/script.js]: Should the rating also be passed in the url?

Comment (Tochukwuibe) [reply src/main/webapp/script.js]: I think if have it set to look in the request body for the rating and the completed flag like so 
```
   boolean completed = Boolean.parseBoolean(request.getParameter("completed"));
   int rating = Integer.parseInt(request.getParameter("rating"));
```

So you should probably go with the body, it should look similar to this (what we did for week 2)

```
    const body = new URLSearchParams();
    body.append('text', text);

    const res = await fetch("ENDPOINT_URL", {
        method: 'POST',
        body
    });
```

Comment (Tochukwuibe) [reply src/main/webapp/script.js]: but with the proper values (rating, completed) of course.

Comment (Tochukwuibe) [reply src/main/webapp/script.js]: Just remember to wrap it all in an async function if you want to use the await keyword.

File: src/main/webapp/script.js [hunk 2]
```diff
@@ -16,13 +16,33 @@ $('#myModal').on('shown.bs.modal', function () {
...
+    const itemId = document.getElementById("feedback-id").innerText;
+    const pathId = document.getElementById("path-id").innerText;
+    const options = document.getElementById("ratingValues");
+    const rating = options.options[options.selectedIndex].value;
+
+    const headers = new Headers();
+    headers.append("rating", rating);
+    //If done is always going to be true, is
+    //it really necessary to be sent?
+    headers.append("done", "true");
```
Comment (phaas) [inline src/main/webapp/script.js]: Headers should represent request options, common ones include 'Content-Type', 'Accept-Encoding', or 'Authorization'. The actual data should not be passed in the headers

Comment (phaas) [reply src/main/webapp/script.js]: `FormData` should go in the `body` attribute:

https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#Body

File: src/main/webapp/templates/path.ftl [hunk 3]
```diff
@@ -42,13 +41,49 @@
...
+                        <div class="modal-header">
+                                <h5 class="modal-title" id="itemFeedbackModal">Item Feedback</h5>
+                                <button type="button" class="close" data-dismiss="modal" aria-label="Close">
+                                        <span aria-hidden="true">&times;</span>
+                                </button>
+                        </div>
+                        <div class="modal-body">
+                                <form id="feedback">
+                                        <h4 id="feedback-title">Item Name</h4>
+                                        <h6 id="feedback-id" value="" hidden><h6>
```
Comment (phaas) [inline src/main/webapp/templates/path.ftl]: change `h6` to `input`:

`<input type="text" value="..." hidden>`

File: src/main/webapp/templates/path.ftl [hunk 4]
```diff
@@ -42,13 +41,49 @@
...
+                <div class="modal-content">
+                        <div class="modal-header">
+                                <h5 class="modal-title" id="itemFeedbackModal">Item Feedback</h5>
+                                <button type="button" class="close" data-dismiss="modal" aria-label="Close">
+                                        <span aria-hidden="true">&times;</span>
+                                </button>
+                        </div>
+                        <div class="modal-body">
+                                <form id="feedback">
+                                        <h4 id="feedback-title">Item Name</h4>
```
Comment (phaas) [inline src/main/webapp/templates/path.ftl:65]: Add:
`<input type="text" hidden value="${path.id}" name="path-id">`

File: src/main/webapp/script.js [hunk 5]
```diff
@@ -16,13 +16,29 @@ $('#myModal').on('shown.bs.modal', function () {
...
 
-function submitFeedback() {
-  // TODO
-}
+async function submitFeedback() {
+    console.log("Sending the submission form info.");
+    const itemId = document.getElementById("feedback-id").innerText;
+    const pathId = document.getElementById("path-id").innerText;
+    const options = document.getElementById("ratingValues");
+    const rating = options.options[options.selectedIndex].value;
```
Comment (phaas) [inline src/main/webapp/script.js]: rating = document.getElementById("ratingValues").value;

File: src/main/webapp/script.js [hunk 6]
```diff
@@ -16,13 +16,29 @@ $('#myModal').on('shown.bs.modal', function () {
...
+
+    const body = new URLSearchParams();
```
Comment (phaas) [inline src/main/webapp/script.js:35]: Consider FormData here?

File: src/main/webapp/templates/path.ftl [hunk 7]
```diff
@@ -1,9 +1,8 @@
 <#include "common/header.ftl">
 
-<#include "feedback.ftl">
-
 <div class="container">
         <h1>${path.name}</h1>
+        <p id="path-id" hidden>${path.id}</p>
```
Comment (phaas) [inline src/main/webapp/templates/path.ftl]: Can remove this if we have an <input> in the form

**Review by phaas — CHANGES_REQUESTED** [review]

**Review by Tochukwuibe — APPROVED** [review]

File: src/main/webapp/script.js [hunk 8]
```diff
@@ -16,13 +16,35 @@ $('#myModal').on('shown.bs.modal', function () {
...
+    const pathId = document.getElementById("path-id").value;
+    const rating = document.getElementById("ratingValues").value;
+
+    console.log("TESTING VALUES: itemId " + itemId +
+    " pathId " + pathId + " rating " + rating);
+
+    //const form = $("#feedback");
+    const FD = new FormData();
+    FD.append("rating", rating);
+    FD.append("done", true);
```
Comment (phaas) [inline src/main/webapp/script.js]: Apparently browsers (at least Chrome) encode this form as "multi-part" data, not the normal way that forms are encoded.

```
[INFO] ------WebKitFormBoundaryO3uYUUmZpMgBBcJK
[INFO] Content-Disposition: form-data; name="rating"
[INFO] 
[INFO] 1
[INFO] ------WebKitFormBoundaryO3uYUUmZpMgBBcJK
[INFO] Content-Disposition: form-data; name="done"
[INFO] 
[INFO] true
[INFO] ------WebKitFormBoundaryO3uYUUmZpMgBBcJK--
[INFO] 2020-04-28 09:09:47.590:WARN:oejs.HttpChannel:qtp107456312-26: /learning-path/23/item/16
```

The normal servlet request.getParameter(..) function doesn't know how to interpret this. ugh.

Comment (phaas) [reply src/main/webapp/script.js]: For comparison, here's what the posted data looks like using `URLSearchParams`:

```
[INFO] {rating=[1],done=[true]}
```

**Review by phaas — APPROVED** [review]

**Pull request merged**